src/pubsub.js

```diff
@@ -50,6 +62,94 @@ class BasicPubSub extends Pubsub {
     })
   }
 
+  /**
+   * Dial a received peer.
+   * @override
+   * @param {PeerInfo} peerInfo The peer being dialed
+   * @param {function} callback
+   *
+   * @returns {void}
+   */
+  _dialPeer (peerInfo, callback) {
+    const onDial = (idB58Str, pubsubStopped, conn) => {
+      // pubsub has been stopped, so we should just bail out
+      if (pubsubStopped) {
+        this.log('pubsub was stopped, not processing dial to %s', idB58Str)
+        return callback()
+      }
+
+      this._onDial(peerInfo, conn, callback)
+    }
+
+    callback = callback || function noop () { }
+    const idB58Str = peerInfo.id.toB58String()
+
+    // If already have a PubSub conn, ignore
+    const peer = this.peers.get(idB58Str)
+    if (peer && peer.isConnected) {
+      return nextTick(() => callback())
+    }
+
+    // If already dialing this peer, ignore
+    if (this._dials.has(idB58Str)) {
+      this.log('already dialing %s, ignoring dial attempt', idB58Str)
+      return nextTick(() => callback())
+    }
+
+    // Verify if is known that the peer does not support Gossipsub
+    let notSupportingGossipsub = peerInfo.protocols.size && !peerInfo.protocols.has(this.multicodec)
+
+    // Define multicodec to use
+    // Should fallback to floodsub if fallback is enabled, protocols were negotiated, and no Gossipsub available
+    let multicodec = this.multicodec
+
+    if (this._options.fallbackToFloodsub && notSupportingGossipsub) {
+      multicodec = floodsubMulticodec
+    }
+
+    this._dials.add(idB58Str)
+
+    this.log('dialing %s %s', multicodec, idB58Str)
+
+    this.libp2p.dialProtocol(peerInfo, multicodec, (err, conn) => {
+      this.log('dial to %s complete', idB58Str)
+
+      // If the dial is not in the set, it means that pubsub has been stopped
+      const pubsubStopped = !this._dials.has(idB58Str)
+      this._dials.delete(idB58Str)
+
+      if (err) {
+        // If previously dialed gossipsub and not supported, try floodsub if enabled fallback
+        if (this._options.fallbackToFloodsub &&
+          multicodec === this.multicodec &&
+          err.code === multistreamConstants.errors.MULTICODEC_NOT_SUPPORTED) {
+          this._dials.add(idB58Str)
+          this.log('dialing %s %s', floodsubMulticodec, idB58Str)
+
+          this.libp2p.dialProtocol(peerInfo, floodsubMulticodec, (err, conn) => {
+            this.log('dial to %s complete', idB58Str)
+
+            // If the dial is not in the set, it means that pubsub has been stopped
+            const pubsubStopped = !this._dials.has(idB58Str)
```

Comment:
This is being checked in multiple places, what is this checking for? Is it checking if pubsub itself has stopped? If so, we should avoid having `isStopped` logic in here, it should fail gracefully. If it's not, there's likely a better, central solution.

Reply:
I agree with you. This is something based on the base implementation of pubsub: [libp2p/js-libp2p-pubsub/src/index.js#L163-L166](https://github.com/libp2p/js-libp2p-pubsub/blob/master/src/index.js#L163-L166).

I will create an issue on `js-libp2p-pubsub` also mentioning the logic here. I believe that the best approach is to leverage the refactors to async, with the abort dial to use it on stop, which will not need this anymore.

Reply:
https://github.com/libp2p/js-libp2p-pubsub/issues/21

Reply:
So `_dialPeer` is a misnomer (we should change this in the future). We're not actually dialing the peer. This is triggered from the `peer:connect` event, which means we have a muxed connection to the given peer when this is called. We're not dialing to the peer, we're just negotiating the pubsub protocol. 

In the event that pubsub is shut down in the process of that happening, the call to `dialProtocol` will fail, because the switch should also be stopped. Unless users need to stopping pubsub independently of libp2p, this shouldn't be needed.

Unless taking the logic out causes test failures I would just remove it. We can always add a more appropriate `isStarted` check via libp2p/js-libp2p-pubsub#21, and use a `super` call to get the behavior in a future release.

Regardless of that, I don't think this is a blocker for moving forward with this, but it does make the code harder to follow.

Additionally, thinking about this a bit more, I think it would be worth creating a utility method on libp2p to handle the multi dial negotiation for you, especially as the list of protocols grows, so we don't end up with this logic replicated everywhere. Something like:
```js
const protocolPreferences = [gossipsubMulticodec, floodsubMulticodec]
libp2p.newStream(peerInfo, protocolPreferences, callback)
// or
libp2p.dialBest(peerInfo, protocolPreferences, callback)
```

Reply:
I like the `dialBest` idea and I will create an issue for that in `js-libp2p`! Other than that, I removed the `pubsubStopped` validation. This should be good to go then

Reply:
FYI, created an issue for simplifying this is Gossipsub in a next PR: https://github.com/libp2p/js-libp2p/issues/376

Reply:
FYI, created an issue to improve this in a next PR: https://github.com/libp2p/js-libp2p/issues/376